x-pack/plugins/apm/public/utils/documentation.ts

```diff
@@ -158,7 +158,7 @@ export const ELASTIC_DOCS = {
 //
 // Helper methods
 //
-function translateAgentName(agentName: string): string {
+function translateAgentName(agentName?: string): string | undefined {
```

Comment:
General comment: Typescript has excellent type inference, and we should take full advantage of it. I don't think adding the return type explicitly gives us any more safety than leaving the compiler to do its job.

Where I think return types make sense is for implicit contracts. Eg. the contract between a Node endpoint and the corresponding method in the client making a Rest request to said endpoint. Here it makes sense to specify a shared interface as return type, since Typescript can't figure out their connection.

Reply:
I don't quite follow this right at the moment, but am happy to learn more about this next week and we can add this to the TS cleanup work we want to do.

Reply:
Consider this example:
```ts
const name: string = 'Søren';
```

The type annotation doesn't add any value, since the type can be inferred. Same with the return type in the above case. The types just add a bit of clutter to the function definition, and makes it harder to change the implementation (since one has to update the return type also).

Reply:
Oh, I just came up with this killer argument that I have to write! :D

The problem with explicit return types (when not necessary) is that they coerce the return type, and thus loose type information. Take this example for instance:

Without an explicit type, the compiler will have the most accurate information to base it's decision on:
```ts
function test() {
  return Math.random() > 0.5 ? 'yes' : 'no';
}

const isMaybe = test() === 'maybe'; // ts error: This condition will always return 'false' since the types '"yes" | "no"' and '"maybe"' have no overlap.
```

Explicitly adding a type could coerce the result, thus not giving us the warning we deserve:
```ts
function test(): string {
  return Math.random() > 0.5 ? 'yes' : 'no';
}

const isMaybe = test() === 'maybe'; // no error
```

Reply:
So here the explicit return type is more generic than the inferred type. Makes sense. When to let TS infer things and when to explicitly spell them out is still super fuzzy to me. Can I summarize this convention as "avoid return types unless the caller is external and the type should then be shared across that boundary (REST, library, etc)"?